app/services/plugin.service.ts

```diff
@@ -142,54 +144,56 @@ export class PluginService {
         this.runInputCode = sandboxService.compileScript('sinap.__program.then((program) => program.run(sinap.__input))');
     }
 
-    public getInterpreter(withGraph: Core.Graph): Promise<Program> {
-        let kind = withGraph.plugin.kind;
-        let context = this.interpreters.get(kind);
-        let graph = new Graph(withGraph);
-        var result: Promise<Program>;
-
-        let processContext = (context: any): Promise<Program> => {
-            context.sinap.__graph = graph;
-            console.log(context);
-            return this.interpretCode.runInContext(context);
-        };
-
-        if (context) {
-            result = processContext(context);
-        } else {
-            result = this.loadPlugin(kind, "./build/plugins/dfa-interpreter.js") // TODO: Put real file in here.
-                .then((cont) => {
-                    context = cont;
-                    this.interpreters.set(kind, context);
-                    return processContext(context);
+    public getInterpreter(graph: Core.Graph): Promise<Program> {
+        // TODO: consider: this could really just be a cast and 
+        // graph.plugin.context
+        return this.getPlugin(graph.plugin.kind).context.then((context) => {
+            context.sinap.__graph = new Graph(graph);
+
+            return this.interpretCode
+                .runInContext(context)
+                .then<Program>((program) => {
+                    return {
+                        run: (input: ProgramInput): Promise<ProgramOutput> => {
+                            context.sinap.__input = input;
+                            // Cast is necessary. 
+                            // I prefer forcing it to be explicit (and thus not any)
+                            return this.runInputCode.runInContext(context) as Promise<ProgramOutput>;
+                        },
+                        compilationMessages: [""]
```

Comment:
We can actually get the compilationMessages since they are being produced by the plugin.

```
var script = this.sandboxService.compileScript('sinap.__program.then((program) => program.compilationMessages');
messagePromise = script.runInContext(context)
```
Not sure exactly how to work around the promise API here, though. Will probably be easier once the plugin API is updated to not require promises of plugin developers.